### ISSUE TITLE
Jade support in webpack releases

### ISSUE THREAD

**Issue by ShadowManu** [angular/angular-cli#1886]
Sorry for not providing template information but I only have my phone available (and there's no critical info either).

Is there some in-the-works support for jade? The related issues/PR 's were closed but no real solution was provided. What can be done to have support for this?

**Comment by ShadowManu** [comment]
So you guys closed #996 and #968 in favor of #40, then closed #40 with no specific ending, then closed the efforts in #1433 and #626 because you changed to webpack and here we are with no answers.

Sorry, I'm not trying to be annoying, specially when I see a lot of hate on recent issues with bugs regarding latest TS/Angular/CLI combinations. I can understand error because we are humans, I can understand time because we need to sleep and have to live, but I do hate purposeful indifference, just because its not your problem. I'm open to work myself on the CLI on a new PR, but I do need guidance, and that really starts at least with a single answer in a timelapse of 10 days. Nevertheless, thank you for your efforts.

**Comment by niklas-dahl** [comment]
This is my current "_solution_":

gulpfile.coffee:
```
let path = require('path');
let gulp = require('gulp');
let pug  = require('gulp-pug');

let pugFiles = './src/**/*.pug';

gulp.task('compilePug', () => {
  gulp.src( pugFiles )
    .pipe( pug().on('error', () => console.log('JADE ERROR')) )
    .pipe( gulp.dest('./src') );
});

gulp.task('default', ['compilePug'], () => {
  gulp.watch(pugFiles, ['compilePug']);
});

```
.vscode/settings.json:
```
{
  "files.exclude": {
      "src/app/**/*.html": {"when": "$(basename).pug"}
  }
}
```

This actually works but it's obviously only a temporary solution..

**Comment by ShadowManu** [comment]
My team is using `pug-cli` with `--watch` at the moment for the same results. However, I would prefer jade to be part of the webpack bundling itself. I'll submit a PR if I have time to figure out how to set it up.

**Comment by theklausster** [comment]
Any news on how to add pug to cli?

I opened a thread on Stockoverflow aswell, hoping for some result 
http://stackoverflow.com/questions/39518500/how-to-add-pug-to-angular-cli

**Comment by ShadowManu** [comment]
Opened a PR to see how this goes.

**Comment by filipesilva** [comment]
Heya @ShadowManu, we're not trying to be indifferent and I'm sorry that the jade support has fallen to the wayside.

I can't say jade/pug support is a big on our radars right now to be honest. If you can get a PR to work I'm happy to review it, but bear in mind that it needs to have tests and documentation.

See https://github.com/angular/angular-cli#development-hints-for-hacking-on-angular-cli for a way to start working with a local copy of the CLI.

The command that runs the e2e suite is `node tests/e2e_runner.js`. Have a look at https://github.com/angular/angular-cli/tree/master/tests/e2e/tests/build/styles for sass/less/styl tests.

You should test that the the correct html is generated, and that base jade/pug functionality (like imports) works.

Regarding documentation, something akin to https://github.com/angular/angular-cli#css-preprocessor-integration.

**Comment by ShadowManu** [comment]
Thanks man. I'll be looking at the tests and a proper similar documentation as soon as I have time.

**Comment by KingsDevelopment** [comment]
Any updates @ShadowManu ?

**Comment by nvrossett** [comment]
@ShadowManu please, you can look too the support to (lost-grid and/or jeet.gs) and rupture in stylus?

**Comment by ShadowManu** [comment]
@KingsDevelopment @filipesilva I've worked a little with some time I had on this, but got blocked on how to make the tests pass (specially on the time they take). I'll have a second try tomorrow.

@nvrossett not that I have knowledge of it. And probably should be on a different PR/issue.

**Comment by TaylorDennisLee** [comment]
I just wanted to throw my support behind @tycho01 's pug-plugin-ng and pug-ng-html-loader. I used them to use pug with the Angular Class seed, and it really cleans up the syntax and makes development with Angular2 and pug very pleasant. I recommend checking it out:

https://github.com/tycho01/pug-plugin-ng
https://github.com/tycho01/pug-ng-html-loader

(pug-ng-html-loader is just a fork of pug-html-loader which allows pug-plugin-ng to work)

**Comment by tycho01** [comment]
@TaylorDennisLee: thanks. note that for Webpack 2 it should work without the additional loader as well now, see the readme.

**Comment by TaylorDennisLee** [comment]
@tycho01 Okay, so now pug-plugin-ng should work with pug-html-loader, as long as you are using Webpack 2?

Also, do you still have to do 

template: require('./myComp.pug')

instead of:

templateUrl: 'myComp.pug'


Thanks,

Taylor

**Comment by tycho01** [comment]
@TaylorDennisLee: yeah; if you're experiencing issues feel free to file them.
It does still need the `require` like that, though it seems `angular-cli` appeared to have addressed this for Sass and co., so I imagine they could for template engines like Pug as well.

**Comment by mbonski** [comment]
@tycho01 can you please provide small guide how to add this loader to angular cli. I'm getting:

> src/app/app.component.pug Unexpected token (1:3) You may need an appropriate loader to handle this file type.

    import { Component } from '@angular/core';
    @Component({
    selector: 'app-root',
     template: require('./app.component.pug'),
     styleUrls: ['app.component.scss']
    })
    export class AppComponent {
      title = 'app works!';
    }

**Comment by tycho01** [comment]
@mbonski: I may have use for further info (where/how you set it, first line of the pug file, whether it worked without it). Could you post an issue with those [here](https://github.com/tycho01/pug-plugin-ng/issues)?

**Comment by mbonski** [comment]
@tycho01 I created one. [reference](https://github.com/tycho01/pug-plugin-ng/issues/2)

**Comment by TaylorDennisLee** [comment]
I'd be interested to see how to do this with just temporary script that runs right before "ng serve" or "ng build" and also what in the angular-cli source code a change would naturally go. @tycho01 do you have a link for how they handled the Sass issue?

**Comment by tycho01** [comment]
@TaylorDennisLee you might consider using a Gulp task for Pug compilation then, preferably along with one of the watchy kind. That's what I've been using, failing a better solution thus far. That could look something like this:
```js
let gulp = require('gulp'),
    gulpWatch = require('gulp-watch'),
    runSequence = require('run-sequence'),
    pug = require('gulp-pug'),
    pug_plugin_ng = require('pug-plugin-ng');

let pug_opts = { doctype: 'html', plugins: [pug_plugin_ng], pretty: true };
let toSrc = gulp.dest((file) => file.base);

gulp.task('pug', () =>
  gulp.src('src/**/*.pug')
  .pipe(pug(pug_opts))
  .pipe(toSrc)
);

gulp.task('watch', [], () => {
  runSequence(
    ['pug'],
    () => {
      gulpWatch('src/**/*.pug', () => gulp.start('pug'));
    }
  );
});
```

As to changes, I linked ShadowManu's and mine in mbonski's issue thread.

On the Sass point, I didn't know yet :), but the magic appears to be located [here](https://github.com/angular/angular-cli/blob/f9a7c01602546dea6c92adb165a5c437689c41b4/packages/webpack/src/loader.ts#L111-L125). If I'm reading that correctly, than it should be automatically adding support for any extension already recognized by webpack. So I may have been wrong there earlier!

**Comment by ShadowManu** [comment]
@TaylorDennisLee @tycho01 why not simply using `pug src --watch & ng serve`? Sure, since they're running simultaneously, webpack generally starts doing 2 builds. But hey, config-wise, you just need to install `pug` and `pug-cli` with npm.

**Comment by tycho01** [comment]
To answer @mbonski's earlier question of a guide on how to get this to work before it's merged, with help from @TaylorDennisLee I got it to work as follows:
```
git clone -b pug https://github.com/zoitravel/angular-cli.git
# alternative with plugin: git clone -b pug-html https://github.com/tycho01/angular-cli.git
cd angular-cli
npm link
cd PATH/TO/MY/PROJECT
npm link angular-cli
# in component: templateUrl: './my_pug_file.pug',
ng serve
```

@ShadowManu: Yeah, under normal circumstances I suppose that would've been the straight-forward way. One complication I faced personally with my plugin was (1) adding configuration, and (2) having to ensure I could add Pug plugins as well (-> functions, therefore can't survive JSON serialization, so couldn't be passed through a CLI). I never really wanted to use a plugin, but that's how it ended up. I'd definitely rather not resort to Gulp either.

**Comment by CarlosV89** [comment]
Has anyone gotten variable interpolation to work with webpack1 and pug-ng-html-loader? For some reason its not being able to access my variables

**Comment by CrystalzWind** [comment]
I found a way to use AngularCLI with Pug using pug-html-loader. After generate new projects using ng new.
Install pug and pug-html-loader packages. npm install --save-dev pug pug-html-loader.
Add pug-html-loader to your project \node_modules\@angular\cli\models\webpack-configs\common.js
{ 
        test: /\.(pug|jade)$/, 
        loader: 'pug-html-loader' 
}
Use .jade template in component.

@Component({
  selector: 'app',
  templateUrl: './app.template.jade'
})

I'v made a demo template here https://github.com/CrystalzWind/AngularCLI-Pug

**Comment by donmahallem** [comment]
Equivalent option to @CrystalzWind 's example is to use pug-loader with apply-loader.

`npm install pug-loader apply-loader`

and piping the output of the pug-loader through the apply loader which simply executes the template function.

` {
     test: /\.pug$/,
     loader: "apply-loader!pug-loader?self"
 }`

**Comment by Aides359** [comment]
Looking at https://github.com/angular/angular-cli/blob/master/packages/%40angular/cli/models/webpack-configs/common.ts#L33 and https://github.com/angular/angular-cli/blob/master/packages/%40angular/cli/models/webpack-configs/common.ts#L104-L112 it seems like giving configuration access to the `extraRules` variable (there seems to be some configuration options through `buildOptions` and `appConfig` already) might be enough to solve this issue.

**Comment by filipesilva** [comment]
I'm sorry to be the bearer of bad news, but jade/pug support will not make it to 1.0.  It's an medium/big feature that we can't spare the time to make sure it work properly. 

I know it looks like it would only be a small change in loaders but that's just not been the case with similar features like sass support.

We're looking at it for 2.0 instead, either as base functionality or as one of the main usecases for addons.

**Comment by tycho01** [comment]
Thanks for the update. If we could help out in the testing department, I'd be glad to.

**Comment by Aides359** [comment]
@filipesilva one of the things that would need some consideration for proper pug support would be how to handle the angular syntax with pug since pug has problems with it (would we want to trade equivalence to pug for syntax convenience).

For reference see tycho01/pug-plugin-ng

**Comment by tycho01** [comment]
On that, fixing Pug to play nice with Angular amounted to [one small addition](https://github.com/pugjs/pug-lexer/pull/69/files#diff-168726dbe96b3ce427e7fedce31bb0bcR1066) if it'd made it into Pug core (-> consider '[' and '(' after whitespace as valid indication a new property has started).
While I feel this makes much sense in the Angular 2+ context, I don't consider it a blocker for initial support, in the sense that it's pretty much backward compatible.
In the Angular context I'd be in favor of `doctype: 'html'` as well, to allow `#myVar` / `md-raised-button` over `#myVar=''` / `md-raised-button=''`.

**Comment by MarkPieszak** [comment]
@filipesilva Couldn't everyone just `ng eject` the webpack configuration and just add .pug support now?

**Comment by filipesilva** [comment]
They could, yes. We'd still like to make it a first class citizen one day.

**Comment by Aides359** [comment]
@MarkPieszak this was my current solution to this problem.
Thanks for mentioning.

**Comment by ausir0726** [comment]
hope some day can support pug in angular-cli

**Comment by Anthonyzou** [comment]
@mbonski 
Your example is really close.
First install `pug` and `pug-loader`
```
import { Component } from '@angular/core';
@Component({
selector: 'app-root',
 template: require('pug-loader!./app.component.pug')(),
 styleUrls: ['app.component.scss']
})
export class AppComponent {
  title = 'app works!';
}
```
The only difference is 
` template: require('./app.component.pug')`
To =>
` template: require('pug-loader!./app.component.pug')()`

**Comment by donmahallem** [comment]
So far I got it working with ng eject BUT this does not work with tests which again does desire an appropiate loader.

@Anthonyzou Your workaround does work for JIT but I don't seem to get it running with AOT.
The error reads as follows:

> Error encountered resolving symbol values statically. Calling function 'template', function calls are not supported. Consider replacing the function or lambda with a reference to an exported function, resolving symbol abc in a/b/c

Following the advice of the error and some sites regarding aot I tried:
`
export function template() {
  return require("pug-loader!./test.component.pug")();
}

@Component({
    template: template()
})`

But it still errors with basicly the same message. Any ideas?

**Comment by Anthonyzou** [comment]
@donmahallem I actually have no idea, actually. This is adapted from a graphql angular2 example where they manually require a graphql file and it works with AOT. It must also depend on how the loader is built.

**Comment by filipesilva** [comment]
Heya all, this isn't a feature that we're looking at adding in the 1.x timeframe, but rather something that would happen as a third party addon using the upcoming addon system for v2+.

**Comment by svallory** [comment]
Guys, I made a pull request to allow tweaking webpack config with a file in the project root. For now, you can use the cli at https://github.com/cashfarm/angular-advanced-cli

Install `pug`, `pug-ng-html-loader` and add a `webpack.config.js` with this to root of your project: 
```
module.exports = {
  module: {
    rules: [
      {
        test: /\.(pug|jade)$/,
        use: ['pug-ng-html-loader']
      }
    ]
  }
}
```
Then your components can use pug
```
@Component({
  selector: 'app-root',
  templateUrl: './app.component.pug',
  styleUrls: ['./app.component.styl']
})
```
But your index has to stay in html.

**Comment by Pexak** [comment]
Any update?

**Comment by MarkPieszak** [comment]
Going along with some of the recommendations above, I created a little script that can be ran to automatically insert the pug rule into the node_modules folder.

This can be used as a temporary solution until it can be available natively. Note this could need to be updated & changed if the CLI changes the location or formatting of this file. It's naively searching for `rules: [` within this specific file.

```bash
npm i --D apply-loader pug-loader
```


### pug-rule-insert.js

Now you can just create a simple js file with the code below (put it at your root project level).

```typescript
const fs = require('fs');

const commonCliConfig = 'node_modules/@angular/cli/models/webpack-configs/common.js';
const pug_rule = `\n{ test: /\.pug$/, loader: "apply-loader!pug-loader?self" },`;

fs.readFile(commonCliConfig, (err, data) => {
  if (err) { throw err; }

  const configText = data.toString();

  if (configText.indexOf(pug_rule) > -1) {
    return;
  }

  console.log('-- Inserting .pug webpack rule -- ');

  const position = configText.indexOf('rules: [') + 8;
  const output = [configText.slice(0, position), pug_rule, configText.slice(position)].join('');

  const file = fs.openSync(commonCliConfig, 'r+');
  fs.writeFile(file, output);
  fs.close(file);
});
```

### Running it

Run it simply with: `node pug-rule-insert.js` or create a nice `postinstall` hook to do it for you on new installs.

`"postinstall": "node pug-rule-insert.js"`

Needed this for a client, hope this helps!

Works with `ng serve` and `AoT` prod builds, and even *Universal + CLI* projects 🎁

**Comment by clovisj** [comment]
Tnks @MarkPieszak

**Comment by chenzhiguang** [comment]
@MarkPieszak thank you. I used pug-html-loader and html-loader instead, it is working fine, including the *include ./pug-file*

`const pug_rule = `\n{ test: /\.pug$/, loader: [ 'html-loader', { loader: "pug-html-loader", options: { doctype: 'html', pretty: true } } ], },`;
`

**Comment by oguimbal** [comment]
@MarkPieszak @clovisj @chenzhiguang 
[Warning] this trick works fine, as long as you do not name your template files something.TS.pug  (which I used to do so they pop below their .ts counterpart in vscode)

If you do so, your AOT builds will fail silently (at least on my machine, as of today) 

This one was f***ng hard to find.  see #8904
